File: codes/two_pointers/string_compression.py
from typing import List


class Solution:
    def compress(self, chars: List[str]) -> int:
        left, right = 0, 0

        while right < len(chars):
            # right: count
            cur = chars[right]
            cnt = 0
            while right < len(chars) and cur == chars[right]:
                right += 1
                cnt += 1

            # left: update the pattern
            chars[left] = cur
            if cnt == 1:
                left += 1
            else:
                left += 1
                cnt_str = str(cnt)
                for i in range(len(cnt_str)):
                    chars[left] = cnt_str[i]
                    left += 1

        return left
        

File: codes/two_pointers/string_compression.cpp
#include "../base.h"

class Solution {
public:
    int compress(vector<char>& chars) {
        int n = chars.size(), cur = 0;
        int i = 0, j = 0;
        while (i < n) {
            while (j < n && chars[j] == chars[i])
                j++;
            chars[cur] = chars[i];
            cur ++;
            if (j - i != 1) {
                for (char c : to_string(j - i))
                    chars[cur ++] = c;
            }
            i = j;
        }

        return cur;
    }
};